### DISCUSSION TITLE
Can anybody has Vertica AdoNet JobStore implementation

### DISCUSSION THREAD

**Discussion by ghost** [quartznet/quartznet#1447] [General] [answered]
Hi Team,

We are using Vertica as database in our application. I could not find a generic Ado.net implementation for Quartz.net. Can anybody help me to create a AdoNet store provider class for Vertica

Thanks,
Jineesh

**Comment by lahma** [comment]
I think you could start by creating your own driver delegate and check how other custom ones have been built.

**Reply by ghost** [reply]
Thanks for the quick response. Is there any similar implementation available for ADO.net? Can I get an example for extending the deligate? 

I tried to create an override for the class StdAdoDelegate but it seems like so many references there. Is there any easy way to do this?

Thanks again,
Jineesh

**Reply by lahma** [reply]
Just check the implementations under AdoJobStore, here's [OracleDelegate](https://github.com/quartznet/quartznet/blob/main/src/Quartz/Impl/AdoJobStore/OracleDelegate.cs) for example which maps boolean values with custom logic.

**Reply by ghost** [reply]
Thanks for the quick oversight. 

In the configuraiton I have noticed the below setting
quartz.dataSource.myDS.provider = MySql

But if I am going to use custom provider(ie, Vertica) how can I make it work

M

**Reply by lahma** [reply]
https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#quartz-datasource-name-provider

https://github.com/quartznet/quartznet/blob/09b33a20ef5e1caf710fa07ac83a17bcfb0ed7bc/src/Quartz.Tests.Unit/SchedulerBuilderTest.cs#L31-L48

**Reply by ghost** [reply]
Thanks for your quick response. 

I think I am getting closer now. I have used the following code with custom delegation and custom provider. But I could not find a method to set the metadata in the interface. 


var properties = new NameValueCollection();
properties["quartz.jobStore.type"] = "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz";
properties["quartz.jobStore.driverDelegateType"] = typeof(VerticaDelegate).AssemblyQualifiedNameWithoutVersion();
properties["quartz.jobStore.dataSource"] = "default";
properties["quartz.jobStore.useProperties"] = false.ToString();
properties["quartz.serializer.type"] = "json";
properties["quartz.dataSource.default.connectionString"] = GetConnectionString();
properties["quartz.dataSource.default.provider"] = typeof(CustomConnectionProvider).AssemblyQualifiedNameWithoutVersion();

StdSchedulerFactory factory = new StdSchedulerFactory(properties);
IScheduler scheduler = await factory.GetScheduler();
await scheduler.Start();

But I am getting following exception 
`
Inner Exception 1:
ArgumentOutOfRangeException: There is no metadata information for provider 'Quartz3.CustomConnectionProvider, Quartz3'
Parameter name: providerName
`

**Reply by ghost** [reply]
Sorry there is a mistake from my side. I put wrong property name for custom conneciton provider. Its working with

properties["quartz.dataSource.default.connectionProvider.type"] = typeof(CustomConnectionProvider).AssemblyQualifiedNameWithoutVersion();

Thanks for your help. Let me extend the provider and see

**Reply by ghost** [reply]
Hi Lahma,
Sorry to disturb you again. I am sttuck with some other issues. I am still getting some other error. 

"Couldn't recover jobs: Object reference not set to an instance of an object."

I have used following code snippet

`// you can have base properties
var properties = new NameValueCollection();
properties["quartz.jobStore.type"] = "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz";
properties["quartz.jobStore.driverDelegateType"] = typeof(VerticaDelegate).AssemblyQualifiedNameWithoutVersion();
properties["quartz.jobStore.dataSource"] = "default";
properties["quartz.jobStore.useProperties"] = false.ToString();
properties["quartz.serializer.type"] = "json";
properties["quartz.dataSource.default.connectionString"] = GetConnectionString();
properties["quartz.dataSource.default.connectionProvider.type"] = typeof(CustomConnectionProvider).AssemblyQualifiedNameWithoutVersion();
StdSchedulerFactory factory = new StdSchedulerFactory(properties);
IScheduler scheduler = await factory.GetScheduler();
await scheduler.Start();
`

For the custom provider, I have used VerticaConnection and VerticaCommand as return types. 

I found another thing. As property file "Quartz.Impl.AdoJobStore.Common.dbproviders.properties" can be seen in the source code. Is there a way to add Vertica database driver information in this file programatically and use a VerticaDeligate 

Please suggest

Jineesh

**Reply by ghost** [reply]
Hi lahma,

I am able to implement this without the propery file. Now everything working except one thing. 

If I put the schema name along with the table prefix the scheduler.Start() method is not loading. 

properties["quartz.jobStore.tablePrefix"] = "myschema.QRTZ_";

I could not find any error as well.  If I comment out the above statement I am able to see the custom deligate overrides got fired. 
Please help.

**Reply by ghost** [reply]
Hi Lahma,

Could you please help me on this? If I specify the tablePrefix in properties I am getting stuck on sheduler.Start(). If I comment out the tablePrefix its getting triggered. See the event log if I put the tablePrefix. No other log after that and applciation simply termintated wihtout generating any error. 

03/Dec/2021 12:22:01 PM [1] [INFO] AddConnectionProvider - Registering datasource 'default' with db provider: 'Quartz3.CustomConnectionProvider'
03/Dec/2021 12:22:01 PM [1] [INFO] MoveNext - Using object serializer: Quartz.Simpl.JsonObjectSerializer, Quartz.Serialization.Json
03/Dec/2021 12:22:01 PM [1] [DEBUG] Initialize - TaskSchedulingThreadPool configured with max concurrency of 10 and TaskScheduler ThreadPoolTaskScheduler.
03/Dec/2021 12:22:01 PM [1] [INFO] .ctor - Initialized Scheduler Signaller of type: Quartz.Core.SchedulerSignalerImpl
03/Dec/2021 12:22:01 PM [1] [INFO] ? - Quartz Scheduler created
03/Dec/2021 12:22:01 PM [1] [DEBUG] AddTriggerPersistenceDelegate - Adding TriggerPersistenceDelegate of type: Quartz.Impl.AdoJobStore.SimpleTriggerPersistenceDelegate
03/Dec/2021 12:22:01 PM [1] [DEBUG] AddTriggerPersistenceDelegate - Adding TriggerPersistenceDelegate of type: Quartz.Impl.AdoJobStore.CronTriggerPersistenceDelegate
03/Dec/2021 12:22:01 PM [1] [DEBUG] AddTriggerPersistenceDelegate - Adding TriggerPersistenceDelegate of type: Quartz.Impl.AdoJobStore.CalendarIntervalTriggerPersistenceDelegate
03/Dec/2021 12:22:01 PM [1] [DEBUG] AddTriggerPersistenceDelegate - Adding TriggerPersistenceDelegate of type: Quartz.Impl.AdoJobStore.DailyTimeIntervalTriggerPersistenceDelegate
03/Dec/2021 12:22:01 PM [1] [INFO] Initialize - Using thread monitor-based data access locking (synchronization).
03/Dec/2021 12:22:01 PM [1] [INFO] Initialize - JobStoreTX initialized.
03/Dec/2021 12:22:01 PM [1] [INFO] Initialize - Scheduler meta-data: Quartz Scheduler (v3.3.3.0) 'TestScheduler' with instanceId 'instance_one'
  Scheduler class: 'Quartz.Core.QuartzScheduler' - running locally.
  NOT STARTED.
  Currently in standby mode.
  Number of jobs executed: 0
  Using thread pool 'Quartz.Simpl.DefaultThreadPool' - with 10 threads.
  Using job-store 'Quartz.Impl.AdoJobStore.JobStoreTX' - which supports persistence. and is not clustered.
03/Dec/2021 12:22:01 PM [1] [INFO] MoveNext - Quartz scheduler 'TestScheduler' initialized
03/Dec/2021 12:22:01 PM [1] [INFO] MoveNext - Quartz scheduler version: 3.3.3.0
03/Dec/2021 12:22:43 PM [1] [DEBUG] MoveNext - Lock 'TRIGGER_ACCESS' is desired by: 66e830e9-8829-4367-815f-258a4f4def9f
03/Dec/2021 12:22:43 PM [1] [DEBUG] MoveNext - Lock 'TRIGGER_ACCESS' is being obtained: 66e830e9-8829-4367-815f-258a4f4def9f
03/Dec/2021 12:22:43 PM [1] [DEBUG] MoveNext - Lock 'TRIGGER_ACCESS' given to: 66e830e9-8829-4367-815f-258a4f4def9f

**Reply by lahma** [reply]
This is weird. Seems that everything is starting as expected and even lock handler is doing its work. Does you example application have a `await Task.Delay()` at the end to ensure process doesn't exist?

**Reply by ghost** [reply]
I am able to make it work by adding overrides for almost all the methods for StdAdoDeligate class. 

So I did following things:
1. Set the property `properties["quartz.jobStore.tablePrefix"] = "[schema].QRTZ_"`;
2. Create a `CustomConnectionProvider : IDbProvider` to override DbComamnd and Connection with Vertica equavalent
3. Create a deleigate `VerticaDelegate : StdAdoDelegate`

Everything seems working fine now. But I think I can minimize the override methods in the VerticaDeligate. `dbProvider.Metadata.DbBinaryType` is the only dependency variable I can see there. Is there a way to assign this globally without overriding the methods? I can see the method `public virtual void Initialize(DelegateInitializationArgs args)`. Is it the best place to build a provider variable with Metadata dynamically?

**Reply by ghost** [reply]
Finally, I have successfully implemented the connection without many changes in the code. 

Attached code snippet. Thanks for your help
[VerticaDeligate_cs.txt](https://github.com/quartznet/quartznet/files/7676800/VerticaDeligate_cs.txt)

**Comment by ghost** [comment, marked as answer]
Finally, I have successfully implemented the connection without many changes in the code. 

Attached code snippet. Thanks for your help
[VerticaDeligate_cs.txt](https://github.com/quartznet/quartznet/files/7676800/VerticaDeligate_cs.txt)

Thanks Jineesh

**Reply by lahma** [reply]
Great to hear you got it working! There's also the `DbProvider.RegisterDbMetadata` which allows you to define a name for provider and the metadata you have like in your code. Might reduce required code even further.

**Reply by ghost** [reply]
I am using the custom provider. So what will be the "providerName" for RegisterDbMetadata ?

Thanks,
Jineesh

**Reply by lahma** [reply]
You could register it with name like "Vertica" and in you Quartz config reference provider with name "Vertica" in props.

**Reply by ghost** [reply]
I have added the line  `properties["quartz.dataSource.default.provider"] = "Vertica"; ` before creating Scheduler `IScheduler quartzScheduler = await SchedulerBuilder.Create(properties).BuildScheduler();` and placed `DbProvider.RegisterDbMetadata`  from DbProvider initialize. But this not working as expected.

**Reply by ghost** [reply]
aha. Got it :)
I am able to skip the CustomProvider class completely. Thanks Sir.

**Reply by lahma** [reply]
Awesome, enjoy!

**Reply by ghost** [reply]
Latest minimal version is attached. 

[VerticaDeligate_cs.txt](https://github.com/quartznet/quartznet/files/7685799/VerticaDeligate_cs.txt)

In order to load quartz.config file properties I am using a function GetConfigurationFileProperties(). Is that required? I can't see the properties loading from quartz.config file if I use `IScheduler quartzScheduler = await SchedulerBuilder.Create(properties).BuildScheduler()`

private NameValueCollection GetConfigurationFileProperties()
{
    string propFileName = FileUtil.ResolveFile("~/quartz.config") ?? "quartz.config";
    if (File.Exists(propFileName))
    {
        // file system
        try
        {
            PropertiesParser pp = PropertiesParser.ReadFromFileResource(propFileName);                    
            Trace.WriteLine($"Quartz.NET properties loaded from configuration file '{propFileName}'");
            return pp.UnderlyingProperties;
        }
        catch (Exception ex)
        {
            throw;                    
        }
    }
    return null;
}

**Reply by lahma** [reply]
Quartz should try to automatically resolve the quartz.config from run directory, you can also set all the properteis in MyApp.exe.config/appSettings.json. See Examples in the Quartz repo.

**Reply by ghost** [reply]
I have tried all different options but nothing works. If I remove the line `NameValueCollection properties = GetConfigurationFileProperties();` nothing will work. I have to add a couple of properties from the code itself so it si required to get the loaded properties in a variable before calling `IScheduler quartzScheduler = await SchedulerBuilder.Create(properties).BuildScheduler()`. Do you have any suggestions?
jineesh

**Reply by lahma** [reply]
Try to debug through Quartz sources and see what gets loaded and why 🙂

**Reply by ghost** [reply]
I have removed `properties` parameter from the `SchedulerBuilder.Create().BuildSheduler()`. And set additional property inside the `.userPerisistantStore()`. Hope that's it. Thanks for your great help. All good for now. 

IScheduler quartzScheduler = await SchedulerBuilder.Create()                
    .UseDefaultThreadPool(x => x.MaxConcurrency = 5)
    .WithMisfireThreshold(TimeSpan.FromSeconds(60))   
    .UsePersistentStore(x =>
    {
        x.UseProperties = true;
        x.UseClustering();                    
        x.UseJsonSerializer();
        x.SetProperty("quartz.jobStore.tablePrefix", string.Format("{0}.QRTZ_", ApplicationConstants.VerticaSchema));
        x.SetProperty("quartz.jobStore.driverDelegateType", typeof(VerticaDelegate).AssemblyQualifiedName);
        x.SetProperty("quartz.jobStore.dataSource", "default");
        x.SetProperty("quartz.dataSource.default.connectionString", ApplicationConstants.VerticaConnectionString);
        x.SetProperty("quartz.dataSource.default.connectionProvider.type", typeof(CustomConnectionProvider).AssemblyQualifiedName);
    })
    .UseXmlSchedulingConfiguration(x =>
    {
        x.Files = new[] { "~/quartz_jobs.xml" };                    
        x.FailOnFileNotFound = true;                    
        x.FailOnSchedulingError = true;
    })
    .BuildScheduler();
